Speed up cincoffset_impl()

arichardson opened the pull request:
Profiling with perf indicates that approximately 10% of the purecap
CHERI-RISC-V boot time is spent in helper_cincoffset. This commit adds
two optimizations:
- Inline the check for the common case (in-bounds update) to avoid a
  function call to the cheri-compressed-cap helpers
- Avoid copying the cap_register_t when source == dest and instead update
  the cr_cursor member directly. While this would also be safe if the
  capability becomes unrepresentable, we only if the result is also
  representable.

This results in a noticeable 1.07x speedup:

```
alr48@waltham:/local/scratch/alr48/cheri/cheribsd(dev u=)> hyperfine -L qemu /local/scratch/alr48/cheri/output/sdk/bin/qemu-system-riscv64cheri.softtlb-new,/local/scratch/alr48/cheri/output/sdk/bin/qemu-system-riscv64cheri  '{qemu} -M virt -m 2048 -nographic -bios bbl-riscv64cheri-virt-fw_jump.bin -kernel /local/scratch/alr48/cheri/output/kernel-riscv64-purecap.CHERI-PURECAP-QEMU-MFS-ROOT -append init_path=/sbin/startup-benchmark.sh' -m 5
Benchmark #1: /local/scratch/alr48/cheri/output/sdk/bin/qemu-system-riscv64cheri.softtlb-new -M virt -m 2048 -nographic -bios bbl-riscv64cheri-virt-fw_jump.bin -kernel /local/scratch/alr48/cheri/output/kernel-riscv64-purecap.CHERI-PURECAP-QEMU-MFS-ROOT -append init_path=/sbin/startup-benchmark.sh
  Time (mean ± σ):     11.107 s ±  0.057 s    [User: 10.727 s, System: 0.125 s]
  Range (min … max):   11.043 s … 11.160 s    5 runs

Benchmark #2: /local/scratch/alr48/cheri/output/sdk/bin/qemu-system-riscv64cheri -M virt -m 2048 -nographic -bios bbl-riscv64cheri-virt-fw_jump.bin -kernel /local/scratch/alr48/cheri/output/kernel-riscv64-purecap.CHERI-PURECAP-QEMU-MFS-ROOT -append init_path=/sbin/startup-benchmark.sh
  Time (mean ± σ):     10.424 s ±  0.162 s    [User: 9.983 s, System: 0.118 s]
  Range (min … max):   10.264 s … 10.621 s    5 runs

Summary
  '/local/scratch/alr48/cheri/output/sdk/bin/qemu-system-riscv64cheri -M virt -m 2048 -nographic -bios bbl-riscv64cheri-virt-fw_jump.bin -kernel /local/scratch/alr48/cheri/output/kernel-riscv64-purecap.CHERI-PURECAP-QEMU-MFS-ROOT -append init_path=/sbin/startup-benchmark.sh' ran
    1.07 ± 0.02 times faster than '/local/scratch/alr48/cheri/output/sdk/bin/qemu-system-riscv64cheri.softtlb-new -M virt -m 2048 -nographic -bios bbl-riscv64cheri-virt-fw_jump.bin -kernel /local/scratch/alr48/cheri/output/kernel-riscv64-purecap.CHERI-PURECAP-QEMU-MFS-ROOT -append init_path=/sbin/startup-benchmark.sh'
```

arichardson committed:
Speed up cincoffset_impl()

Profiling with perf indicates that approximately 10% of the purecap
CHERI-RISC-V boot time is spent in helper_cincoffset. This commit adds
two optimizations:
- Inline the check for the common case (in-bounds update) to avoid a
  function call to the cheri-compressed-cap helpers
- Avoid copying the cap_register_t when source == dest and instead update
  the cr_cursor member directly. While this would also be safe if the
  capability becomes unrepresentable, we only if the result is also
  representable.

This results in a noticeable 1.07x speedup:

```
alr48@waltham:/local/scratch/alr48/cheri/cheribsd(dev u=)> hyperfine -L qemu /local/scratch/alr48/cheri/output/sdk/bin/qemu-system-riscv64cheri.softtlb-new,/local/scratch/alr48/cheri/output/sdk/bin/qemu-system-riscv64cheri  '{qemu} -M virt -m 2048 -nographic -bios bbl-riscv64cheri-virt-fw_jump.bin -kernel /local/scratch/alr48/cheri/output/kernel-riscv64-purecap.CHERI-PURECAP-QEMU-MFS-ROOT -append init_path=/sbin/startup-benchmark.sh' -m 5
Benchmark #1: /local/scratch/alr48/cheri/output/sdk/bin/qemu-system-riscv64cheri.softtlb-new -M virt -m 2048 -nographic -bios bbl-riscv64cheri-virt-fw_jump.bin -kernel /local/scratch/alr48/cheri/output/kernel-riscv64-purecap.CHERI-PURECAP-QEMU-MFS-ROOT -append init_path=/sbin/startup-benchmark.sh
  Time (mean ± σ):     11.107 s ±  0.057 s    [User: 10.727 s, System: 0.125 s]
  Range (min … max):   11.043 s … 11.160 s    5 runs

Benchmark #2: /local/scratch/alr48/cheri/output/sdk/bin/qemu-system-riscv64cheri -M virt -m 2048 -nographic -bios bbl-riscv64cheri-virt-fw_jump.bin -kernel /local/scratch/alr48/cheri/output/kernel-riscv64-purecap.CHERI-PURECAP-QEMU-MFS-ROOT -append init_path=/sbin/startup-benchmark.sh
  Time (mean ± σ):     10.424 s ±  0.162 s    [User: 9.983 s, System: 0.118 s]
  Range (min … max):   10.264 s … 10.621 s    5 runs

Summary
  '/local/scratch/alr48/cheri/output/sdk/bin/qemu-system-riscv64cheri -M virt -m 2048 -nographic -bios bbl-riscv64cheri-virt-fw_jump.bin -kernel /local/scratch/alr48/cheri/output/kernel-riscv64-purecap.CHERI-PURECAP-QEMU-MFS-ROOT -append init_path=/sbin/startup-benchmark.sh' ran
    1.07 ± 0.02 times faster than '/local/scratch/alr48/cheri/output/sdk/bin/qemu-system-riscv64cheri.softtlb-new -M virt -m 2048 -nographic -bios bbl-riscv64cheri-virt-fw_jump.bin -kernel /local/scratch/alr48/cheri/output/kernel-riscv64-purecap.CHERI-PURECAP-QEMU-MFS-ROOT -append init_path=/sbin/startup-benchmark.sh'
```

nwf commented on target/cheri-common/cheri-lazy-capregs.h:
```diff
@@ -236,6 +236,33 @@ static inline void update_capreg(CPUArchState *env, unsigned regnum,
+{
+    /*
+     * This function can be called to avoid copying a full cap_register_t when
+     * updating a capability register inplace.
```
Maybe move this above the function?  It's also probably good to give some more motivation, and also say something like "`new_cursor` must be representable given the bounds of `source_cap`; see `addr_in_cap_bounds` and `is_representable_cap_with_addr`."

nwf commented on target/cheri-common/op_helper_cheri_common.c:
```diff
@@ -690,37 +690,45 @@ DEFINE_CHERI_STAT(csetaddr);
-                            struct oob_stats_info *oob_info)
+static inline QEMU_ALWAYS_INLINE void
+cincoffset_impl(CPUArchState *env, uint32_t cd, uint32_t cb, target_ulong rt,
+                uintptr_t retpc, struct oob_stats_info *oob_info)
```
That `#else`/`#endif` mid-way through this definition feels ugly.  Since you're in the neighborhood, maybe do away with it and just `#ifdef`-guard the `->num_uses++`?

jrtc27 replied:
Yeah, `struct oob_stats_info *oob_info` will implicitly declare the struct, it's only an error if you attempt to dereference the incomplete type.

jrtc27 replied:
Having said that, that's not new in this code :)

arichardson replied:
Sounds like a good suggestions, I'll do that as a follow-up cleanup.

nwf reviewed (changes requested):
Another good find.  Some nits, but nothing major.

arichardson committed:
Simplify the #ifdef around cincoffset_impl()

Using an incomplete type as the arugment is perfectly fine.
Suggested by @nwf and @jrtc27 in #153.
